File: packages/antgrid_relay_client/test/flow_test.dart
// The flow-control constants are hand-mirrored from
// `packages/antgrid-wire/src/flow.ts`, so nothing but these relationships
// catches a mirror that drifted into an unworkable shape. Each one is a rule
// the sender and receiver both depend on, not a restatement of a literal.
import 'dart:typed_data';

import 'package:antgrid_relay_client/antgrid_relay_client.dart';
import 'package:test/test.dart';

void main() {
  test('a receiver credits at least twice per window', () {
    // At window/2 or coarser the sender would spend most of a window blocked
    // waiting for the credit that releases it.
    expect(kCreditBatchBytes * 2, lessThanOrEqualTo(kChannelWindowBytes));
  });

  test('one maximal frame fits in an empty window', () {
    expect(
      kChannelWindowBytes,
      greaterThanOrEqualTo(kMaxFramePayload + kSealOverheadBytes),
    );
  });

  test('the socket cap leaves headroom above a full channel window', () {
    // Equal limits would let one saturated channel block the other, which is
    // the head-of-line the per-channel window exists to prevent.
    expect(kSocketInflightBytes, greaterThan(kChannelWindowBytes));
  });

  test('the send queue can hold a maximal transfer', () {
    // A message the fragmenter accepts must never be refused by the queue: it
    // would be dropped whole after the caller was told nothing.
    expect(kMaxSendQueueBytes, greaterThanOrEqualTo(kMaxTransferBytes));
  });

  test('the seal overhead matches what the transport actually adds', () async {
    // The sender checks its window BEFORE sealing, so this number is the whole
    // basis of the check. Measured, not asserted against a literal.
    final t = E2eTransportDart(
      sendKey: Uint8List(32)..fillRange(0, 32, 7),
      recvKey: Uint8List(32)..fillRange(0, 32, 9),
    );
    const plaintext = 'flow-control overhead probe';
    final sealed = await t.seal(plaintext);
    expect(sealed.length - utf8ByteLength(plaintext), kSealOverheadBytes);
  });
}

File: packages/antgrid-wire/tests/flow.test.ts
import { describe, it, expect } from "bun:test";
import {
  CHANNEL_WINDOW_BYTES,
  SOCKET_INFLIGHT_BYTES,
  CREDIT_BATCH_BYTES,
  MAX_SEND_QUEUE_BYTES,
  WINDOW_RESYNC_AGE_MS,
  WINDOW_STALL_WARN_MS,
  MAX_FRAME_PAYLOAD,
  MAX_TRANSFER_BYTES,
} from "../src/index";

describe("flow-control constants", () => {
  it("leaves the sender at least three quarters of a window between credits", () => {
    expect(CREDIT_BATCH_BYTES * 2).toBeLessThanOrEqual(CHANNEL_WINDOW_BYTES);
  });

  it("holds one maximal fragment, so a lone big frame never deadlocks a channel", () => {
    expect(CHANNEL_WINDOW_BYTES).toBeGreaterThanOrEqual(MAX_FRAME_PAYLOAD);
  });

  it("leaves headroom for control while one channel's window is full", () => {
    expect(SOCKET_INFLIGHT_BYTES).toBeGreaterThan(CHANNEL_WINDOW_BYTES);
  });

  it("queues at least one maximal transfer before dropping messages whole", () => {
    expect(MAX_SEND_QUEUE_BYTES).toBeGreaterThanOrEqual(MAX_TRANSFER_BYTES);
  });

  it("pins the values the Dart client mirrors by hand", () => {
    expect(CHANNEL_WINDOW_BYTES).toBe(2_097_152);
    expect(SOCKET_INFLIGHT_BYTES).toBe(3_145_728);
    expect(CREDIT_BATCH_BYTES).toBe(524_288);
    expect(WINDOW_RESYNC_AGE_MS).toBe(40_000);
    expect(MAX_SEND_QUEUE_BYTES).toBe(67_108_864);
    expect(WINDOW_STALL_WARN_MS).toBe(5_000);
  });
});
